File: apps/api/src/main/scala/momo/api/domain/Pagination.scala
package momo.api.domain

final case class PageRequest(page: Int, pageSize: Int):
  val offset: Long = (page.toLong - 1L) * pageSize.toLong

final case class PagedResult[A](items: List[A], page: PageRequest, totalItems: Int):
  val totalPages: Int =
    if totalItems <= 0 then 0
    else ((totalItems.toLong + page.pageSize.toLong - 1L) / page.pageSize.toLong).toInt

  val hasPreviousPage: Boolean = page.page > 1 && totalPages > 0
  val hasNextPage: Boolean = totalPages > 0 && page.page < totalPages

File: apps/web/src/shared/lib/pagination.ts
export type PaginationState = {
  hasNextPage: boolean;
  hasPreviousPage: boolean;
  page: number;
  pageSize: number;
  totalItems: number;
  totalPages: number;
};

/** Formats the rendered item range once so every pagination surface uses the same scope text. */
export function formatPaginationRange(pagination: PaginationState): string {
  if (pagination.totalItems === 0) {
    return "0件 / 全0件";
  }
  const start = (pagination.page - 1) * pagination.pageSize + 1;
  const end = Math.min(pagination.page * pagination.pageSize, pagination.totalItems);
  return `${start.toLocaleString()}-${end.toLocaleString()}件 / 全${pagination.totalItems.toLocaleString()}件`;
}
